### PR TITLE
Pipelines API review changes

### PR DESCRIPTION
**Pull request by pakrym**
- Use IMemoryList and relative indexes indide ReadCursor. I temporarily copied `IMemoryList<T>`  to pipelines to remove some members and add another one.
- Extract `IAwaiter<T>` and `Awaitable` into `System.Threading` namespace
- Merge `BufferEnumerator.MoveNext` with `ReadCursor.TryGetBuffer` logic
- Move methods from `ReadCursor` to `ReadableBuffer`.
- Remove `Append` and `Preserve` 
- Return `ReadOnlyMemory` from `RB`
- Remove extra extension methods for write, keep only one that takes `Span` and move it to `WB`

### CODE REVIEW

File: src/System.IO.Pipelines/System/IO/Pipelines/BufferSegment.cs [hunk 1]
```diff
@@ -7,21 +7,40 @@
 
 namespace System.IO.Pipelines
 {
-    internal class BufferSegment
+    public interface IMemoryList<T>
+    {
+        Memory<T> Memory { get; }
+
+        IMemoryList<T> Rest { get; }
```
Comment (pakrym) [inline src/System.IO.Pipelines/System/IO/Pipelines/BufferSegment.cs]: I would also like to rename `Rest` to `Next` it's more in the spirit of linked lists.

**Comment by davidfowl** [comment]
Can you write a summary of all the changes in the PR title... this change is massive and does more than just add the IMemoryList

**Comment by pakrym** [comment]
@davidfowl , I updated the issue, this PR still needs some love in details.

**Comment by KrzysztofCwalina** [comment]
cc: @joshfree, @ahsonkhan, @terrajobst

File: src/System.IO.Pipelines.Networking.Sockets/SocketConnection.cs [hunk 2]
```diff
@@ -598,13 +599,14 @@ private void RecycleSmallBuffer(ref ArraySegment<byte> buffer)
         }
 
         // unsafe+async not good friends
-        private unsafe void SetBuffer(Memory<byte> memory, SocketAsyncEventArgs args, int ignore = 0)
+        private void SetBuffer(ReadOnlyMemory<byte> readOnlyMemory, SocketAsyncEventArgs args, int ignore = 0)
         {
-            if (!memory.TryGetArray(out ArraySegment<byte> segment))
+            var memory = Unsafe.As<ReadOnlyMemory<byte>, Memory<byte>>(ref readOnlyMemory);
```
Comment (KrzysztofCwalina) [inline src/System.IO.Pipelines.Networking.Sockets/SocketConnection.cs]: Take a look at https://github.com/dotnet/corefx/blob/master/src/System.Memory/ref/System.Memory.cs#L437

File: src/System.IO.Pipelines/System/IO/Pipelines/BufferEnumerator.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Buffers;
+
+namespace System.IO.Pipelines
+{
+    /// <summary>
+    /// An enumerator over the <see cref="ReadableBuffer"/>
+    /// </summary>
+    public struct BufferEnumerator
```
Comment (KrzysztofCwalina) [inline src/System.IO.Pipelines/System/IO/Pipelines/BufferEnumerator.cs]: Why not nested type in ReadableBuffer?

File: src/System.IO.Pipelines/System/IO/Pipelines/BufferSegment.cs [hunk 4]
```diff
@@ -7,41 +7,60 @@
 
 namespace System.IO.Pipelines
 {
-    internal class BufferSegment
+    public interface IMemoryList<T>
+    {
+        ReadOnlyMemory<T> Memory { get; }
```
Comment (KrzysztofCwalina) [inline src/System.IO.Pipelines/System/IO/Pipelines/BufferSegment.cs]: Then it should be called IReadOnlyMemoryList, but I thought we can use use R/W list (as we discussed)

File: samples/System.IO.Pipelines.Samples/HttpServer/HttpRequestParser.cs [hunk 5]
```diff
@@ -11,13 +11,13 @@ public class HttpRequestParser
...
-        private PreservedBuffer _path;
-        private PreservedBuffer _method;
+        private byte[] _httpVersion;
+        private byte[] _path;
+        private byte[] _method;
 
-        public ReadableBuffer HttpVersion => _httpVersion.Buffer;
-        public ReadableBuffer Path => _path.Buffer;
-        public ReadableBuffer Method => _method.Buffer;
+        public ReadableBuffer HttpVersion => ReadableBuffer.Create(_httpVersion);
```
[the remark this answers is not in the document]
Comment (pakrym) [reply samples/System.IO.Pipelines.Samples/HttpServer/HttpRequestParser.cs]: Old sample code didn't want to spend to much time churning it.

**Comment by davidfowl** [comment]
How sure are we that we won’t break append when we do want to bring it back?

File: src/System.IO.Pipelines.Extensions/StreamExtensions.cs [hunk 6]
```diff
@@ -59,19 +59,11 @@ public static Task CopyToAsync(this ReadableBuffer buffer, Stream stream)
...
-            }
-            else
-            {
-                // Copy required
-                var array = memory.Span.ToArray();
-                await stream.WriteAsync(array, 0, array.Length).ConfigureAwait(continueOnCapturedContext: false);
-            }
+            // Copy required
+            var array = memory.Span.ToArray();
+            await stream.WriteAsync(array, 0, array.Length).ConfigureAwait(continueOnCapturedContext: false);
```
Comment (davidfowl) [inline src/System.IO.Pipelines.Extensions/StreamExtensions.cs]: Why?

File: src/System.IO.Pipelines.Networking.Sockets/SocketConnection.cs (hunk 2 above)
Comment (davidfowl) [reply src/System.IO.Pipelines.Networking.Sockets/SocketConnection.cs]: I don’t like this change.

**Comment by pakrym** [comment]
@davidfowl we'll add a ton of tests and start using it somewhere to make sure it actually works.

File: src/System.IO.Pipelines/System/IO/Pipelines/ReadableBuffer.cs [hunk 7]
```diff
@@ -276,10 +262,205 @@ public static PreservedBuffer Create(OwnedMemory<byte> data, int offset, int len
...
             }
 
-            var segment = new BufferSegment();
-            segment.SetMemory(data, offset, offset + length);
+            return new ReadableBuffer(data, offset, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static long GetLength(ReadCursor begin, ReadCursor end)
+        {
```
Comment (davidfowl) [inline src/System.IO.Pipelines/System/IO/Pipelines/ReadableBuffer.cs]: Move these to another class. ReadCursorOperations

File: src/System.IO.Pipelines/System/IO/Pipelines/WritableBuffer.cs [hunk 8]
```diff
@@ -76,9 +67,52 @@ public void Commit()
...
         {
             return Pipe.FlushAsync(cancellationToken);
         }
+
+        /// <summary>
+        /// Writes the source <see cref="ReadOnlySpan{Byte}"/> to the <see cref="WritableBuffer"/>.
+        /// </summary>
+        /// <param name="buffer">The <see cref="WritableBuffer"/></param>
+        /// <param name="source">The <see cref="ReadOnlySpan{Byte}"/> to write</param>
+        public void Write(ReadOnlySpan<byte> source)
```
Comment (davidfowl) [inline src/System.IO.Pipelines/System/IO/Pipelines/WritableBuffer.cs]: Why isn’t this an extension method?

File: src/System.IO.Pipelines/System/Threading/Awaitable.cs [hunk 9]
```diff
@@ -3,25 +3,25 @@
 
 using System.Runtime.CompilerServices;
 
-namespace System.IO.Pipelines
+namespace System.Threading
 {
```
Comment (davidfowl) [inline src/System.IO.Pipelines/System/Threading/Awaitable.cs]: This is happening?

Comment (davidfowl) [reply src/System.IO.Pipelines/System/Threading/Awaitable.cs]: Isn’t this supposed to be ValueAwaiter?

File: src/System.IO.Pipelines/System/IO/Pipelines/WritableBuffer.cs (hunk 8 above)
Comment (pakrym) [reply src/System.IO.Pipelines/System/IO/Pipelines/WritableBuffer.cs]: @KrzysztofCwalina want's fewer classes and extension methods.

File: src/System.IO.Pipelines/System/Threading/Awaitable.cs (hunk 9 above)
Comment (pakrym) [reply src/System.IO.Pipelines/System/Threading/Awaitable.cs]: @KrzysztofCwalina want's to keep generalized version in Pipelines if it's no going to BCL.

File: src/System.IO.Pipelines.Extensions/StreamExtensions.cs (hunk 6 above)
Comment (pakrym) [reply src/System.IO.Pipelines.Extensions/StreamExtensions.cs]: Fixed.

File: src/System.IO.Pipelines/System/IO/Pipelines/BufferEnumerator.cs (hunk 3 above)
Comment (pakrym) [reply src/System.IO.Pipelines/System/IO/Pipelines/BufferEnumerator.cs]: It's used in some other places where it's created using cursors only. I'll change it in next iteration.

**Comment by pakrym** [comment]
@davidfowl I've addressed your comments

**Comment by davidfowl** [comment]
Don’t kill the performance while I’m on the beach 😬

**Pull request merged**